From 871a7b8174229f4f876d14511b9818ca20228777 Mon Sep 17 00:00:00 2001
From: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
Date: Mon, 29 Dec 2025 00:00:00 +0000
Subject: [PATCH 1/6] luci-app-upnp: improve existing UI slightly

- Change `Active Service Port Maps` to `Active Port Maps`, use the same
  wording for the table headings and ACL as on the overview status page
- Change description field header to `Added via / description`, always
  include the protocol and clearer/less redundant protocol labels
- Disable IPv6 mapping (`ipv6_disable`): UI option added, UCI exists
- Don't clear dependent options if UPnP IGD/STUN temporarily disabled
- Set `notify_interval` minimum to 900 s (default), as recommended by
  [UDA 1.1] (2x=1800 in the standard), because daemon/OpenWrt wrongly
  suggested 30x less in the past, and to reduce multicast traffic and
  power consumption in wireless networks, clearer help
- Report system instead of service uptime, UUID and service lease file:
  UI options removed (disabled, to keep translations) as rarely used
- Slightly improved some text strings
- Show hint if no suitable configuration found (missing related update),
  link to package manager, and show form as read-only to prevent changes
- Configure service lease file by default in rpcd ucode for full UI
  functionality without option set
- Refactoring by moving ubus connection as jow-/ucode@a58fe47 was merged
- Adapt to JavaScript ES6 and remove some line breaks and format files

Pre-update to comming PR: https://redirect.github.com/openwrt/luci/pull/7822

[UDA 1.1]: https://upnp.org/specs/arch/UPnP-arch-DeviceArchitecture-v1.1.pdf#page=30

Signed-off-by: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
---
 .../resources/view/status/include/80_upnp.js  |  29 ++---
 .../luci-static/resources/view/upnp/upnp.js   | 107 +++++++++++-------
 .../usr/share/rpcd/acl.d/luci-app-upnp.json   |   2 +-
 .../root/usr/share/rpcd/ucode/luci.upnp       |  56 ++++-----
 4 files changed, 104 insertions(+), 90 deletions(-)

diff --git a/applications/luci-app-upnp/htdocs/luci-static/resources/view/status/include/80_upnp.js b/applications/luci-app-upnp/htdocs/luci-static/resources/view/status/include/80_upnp.js
index 67e255caea7a..45f03c68a457 100644
--- a/applications/luci-app-upnp/htdocs/luci-static/resources/view/status/include/80_upnp.js
+++ b/applications/luci-app-upnp/htdocs/luci-static/resources/view/status/include/80_upnp.js
@@ -7,14 +7,14 @@
 const callUpnpGetStatus = rpc.declare({
 	object: 'luci.upnp',
 	method: 'get_status',
-	expect: {  }
+	expect: {}
 });
 
 const callUpnpDeleteRule = rpc.declare({
 	object: 'luci.upnp',
 	method: 'delete_rule',
-	params: [ 'token' ],
-	expect: { result : "OK" },
+	params: ['token'],
+	expect: { result: 'OK' },
 });
 
 function handleDelRule(num, ev) {
@@ -35,23 +35,23 @@ return baseclass.extend({
 	},
 
 	render: function(data) {
-		var table = E('table', { 'class': 'table', 'id': 'upnp_status_table' }, [
+		const table = E('table', { 'class': 'table', 'id': 'upnp_status_table' }, [
 			E('tr', { 'class': 'tr table-titles' }, [
-				E('th', { 'class': 'th' }, _('Client Name')),
-				E('th', { 'class': 'th' }, _('Client Address')),
-				E('th', { 'class': 'th' }, _('Client Port')),
-				E('th', { 'class': 'th' }, _('External Port')),
+				E('th', { 'class': 'th' }, _('Hostname')),
+				E('th', { 'class': 'th' }, _('IP address')),
+				E('th', { 'class': 'th' }, _('Port')),
+				E('th', { 'class': 'th' }, _('External port')),
 				E('th', { 'class': 'th' }, _('Protocol')),
 				E('th', { 'class': 'th right' }, _('Expires')),
-				E('th', { 'class': 'th' }, _('Description')),
+				E('th', { 'class': 'th' }, _('Added via / description')),
 				E('th', { 'class': 'th cbi-section-actions' }, '')
 			])
 		]);
 
-		var rules = Array.isArray(data[0].rules) ? data[0].rules : [];
+		const rules = Array.isArray(data[0].rules) ? data[0].rules : [];
 
-		var rows = rules.map(function(rule) {
-			const padnum = (num, length) => num.toString().padStart(length, "0");
+		const rows = rules.map(function(rule) {
+			const padnum = (num, length) => num.toString().padStart(length, '0');
 			const expires_sec = rule?.expires || 0;
 			const hour = Math.floor(expires_sec / 3600);
 			const minute = Math.floor((expires_sec % 3600) / 60);
@@ -72,8 +72,9 @@ return baseclass.extend({
 				rule.descr,
 				E('button', {
 					'class': 'btn cbi-button-remove',
-					'click': L.bind(handleDelRule, this, rule.num)
-				}, [ _('Delete') ])
+					'click': L.bind(handleDelRule, this, rule.num),
+					'title': _('Delete')
+				}, [_('Delete')])
 			];
 		});
 
diff --git a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
index 4657113b5813..5b664db4c5be 100644
--- a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
+++ b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
@@ -3,27 +3,28 @@
 'require dom';
 'require poll';
 'require uci';
+'require ui';
 'require rpc';
 'require form';
 
 const callInitAction = rpc.declare({
 	object: 'luci',
 	method: 'setInitAction',
-	params: [ 'name', 'action' ],
+	params: ['name', 'action'],
 	expect: { result: false }
 });
 
 const callUpnpGetStatus = rpc.declare({
 	object: 'luci.upnp',
 	method: 'get_status',
-	expect: {  }
+	expect: {}
 });
 
 const callUpnpDeleteRule = rpc.declare({
 	object: 'luci.upnp',
 	method: 'delete_rule',
-	params: [ 'token' ],
-	expect: { result : "OK" },
+	params: ['token'],
+	expect: { result: 'OK' },
 });
 
 function handleDelRule(num, ev) {
@@ -44,10 +45,10 @@ return view.extend({
 
 	poll_status: function(nodes, data) {
 
-		var rules = Array.isArray(data[0].rules) ? data[0].rules : [];
+		const rules = Array.isArray(data[0].rules) ? data[0].rules : [];
 
-		var rows = rules.map(function(rule) {
-			const padnum = (num, length) => num.toString().padStart(length, "0");
+		const rows = rules.map(function(rule) {
+			const padnum = (num, length) => num.toString().padStart(length, '0');
 			const expires_sec = rule?.expires || 0;
 			const hour = Math.floor(expires_sec / 3600);
 			const minute = Math.floor((expires_sec % 3600) / 60);
@@ -68,8 +69,9 @@ return view.extend({
 				rule.descr,
 				E('button', {
 					'class': 'btn cbi-button-remove',
-					'click': L.bind(handleDelRule, this, rule.num)
-				}, [ _('Delete') ])
+					'click': L.bind(handleDelRule, this, rule.num),
+					'title': _('Delete')
+				}, [_('Delete')])
 			];
 		});
 
@@ -80,35 +82,42 @@ return view.extend({
 
 		let m, s, o;
 
-		var protocols = '%s & %s/%s'.format(
+		const protocols = _('%s & %s/%s', '%s & %s/%s (%s = UPnP IGD, %s = PCP, %s = NAT-PMP)').format(
 			'<a href="https://en.wikipedia.org/wiki/Internet_Gateway_Device_Protocol" target="_blank" rel="noreferrer"><abbr title="UPnP Internet Gateway Device (Control Protocol)">UPnP IGD</abbr></a>',
 			'<a href="https://en.wikipedia.org/wiki/Port_Control_Protocol" target="_blank" rel="noreferrer"><abbr title="Port Control Protocol">PCP</abbr></a>',
 			'<a href="https://en.wikipedia.org/wiki/NAT_Port_Mapping_Protocol" target="_blank" rel="noreferrer"><abbr title="NAT Port Mapping Protocol">NAT-PMP</abbr></a>');
-		m = new form.Map('upnpd', [_('UPnP IGD & PCP/NAT-PMP Service')],
+		m = new form.Map('upnpd', _('UPnP IGD & PCP/NAT-PMP Service'),
 			_('The %s protocols allow clients on the local network to configure port maps/forwards on the router autonomously.',
 				'The %s (%s = UPnP IGD & PCP/NAT-PMP) protocols allow clients on the local network to configure port maps/forwards on the router autonomously.')
-				.format(protocols)
+			.format(protocols)
 		);
+		if (!uci.get('upnpd', 'config')) {
+			ui.addNotification(null, E('div', '<h4>' + _('No suitable configuration was found!') + '</h4><p>' +
+				_('No suitable (LuCI app %s) config found in %s. Related package update (daemon or LuCI app) may be missing.').format('v1.0', '<code>/etc/config/upnpd</code>') + '<br>' +
+				_('Use the software package manager, update lists, and install the related update. Config is migrated on the daemon package update.') + '</p>' +
+				'<a class="btn" href="/cgi-bin/luci/admin/system/package-manager?query=UPnP%20IGD%20&%20PCP/NAT-PMP">' + _('Go to package manager…') + '</a>'), 'warning');
+			m.readonly = true;
+		}
 
 		s = m.section(form.GridSection, '_active_rules');
 
 		s.render = L.bind(function(view, section_id) {
-			var table = E('table', { 'class': 'table cbi-section-table', 'id': 'upnp_status_table' }, [
+			const table = E('table', { 'class': 'table cbi-section-table', 'id': 'upnp_status_table' }, [
 				E('tr', { 'class': 'tr table-titles' }, [
-					E('th', { 'class': 'th' }, _('Client Name')),
-					E('th', { 'class': 'th' }, _('Client Address')),
-					E('th', { 'class': 'th' }, _('Client Port')),
-					E('th', { 'class': 'th' }, _('External Port')),
+					E('th', { 'class': 'th' }, _('Hostname')),
+					E('th', { 'class': 'th' }, _('IP address')),
+					E('th', { 'class': 'th' }, _('Port')),
+					E('th', { 'class': 'th' }, _('External port')),
 					E('th', { 'class': 'th' }, _('Protocol')),
 					E('th', { 'class': 'th right' }, _('Expires')),
-					E('th', { 'class': 'th' }, _('Description')),
+					E('th', { 'class': 'th' }, _('Added via / description')),
 					E('th', { 'class': 'th cbi-section-actions' }, '')
 				])
 			]);
 
-			var rules = Array.isArray(data[0].rules) ? data[0].rules : [];
+			const rules = Array.isArray(data[0].rules) ? data[0].rules : [];
 
-			var rows = rules.map(function(rule) {
+			const rows = rules.map(function(rule) {
 				return [
 					rule.host_hint || _('Unknown'),
 					rule.intaddr,
@@ -118,15 +127,17 @@ return view.extend({
 					rule.descr,
 					E('button', {
 						'class': 'btn cbi-button-remove',
-						'click': L.bind(handleDelRule, this, rule.num)
-					}, [ _('Delete') ])
+						'click': L.bind(handleDelRule, this, rule.num),
+						'title': _('Delete')
+					}, [_('Delete')])
 				];
 			});
 
 			cbi_update_table(table, rows, E('em', _('There are no active port maps.')));
 
 			return E('div', { 'class': 'cbi-section cbi-tblsection' }, [
-					E('h3', _('Active Service Port Maps')), table ]);
+				E('h3', _('Active Port Maps')), table
+			]);
 		}, o, this);
 
 		s = m.section(form.NamedSection, 'config', 'upnpd', _('Service Settings'));
@@ -135,7 +146,7 @@ return view.extend({
 		s.tab('advanced', _('Advanced Settings'));
 
 		o = s.taboption('setup', form.Flag, 'enabled', _('Start service'),
-			_('Start autonomous port mapping service'));
+			_('Start the autonomous port mapping service'));
 		o.rmempty = false;
 
 		o = s.taboption('setup', form.Flag, 'enable_upnp', _('Enable UPnP IGD protocol'));
@@ -149,86 +160,104 @@ return view.extend({
 		o.default = '1';
 		o.rmempty = false;
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
 		o = s.taboption('setup', form.Value, 'download', _('Download speed'),
 			_('Report maximum download speed in kByte/s'));
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
 		o = s.taboption('setup', form.Value, 'upload', _('Upload speed'),
 			_('Report maximum upload speed in kByte/s'));
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
 		s.taboption('advanced', form.Flag, 'use_stun', _('Use %s', 'Use %s (%s = STUN)')
-				.format('<a href="https://en.wikipedia.org/wiki/STUN" target="_blank" rel="noreferrer"><abbr title="Session Traversal Utilities for NAT">STUN</abbr></a>'),
+			.format('<a href="https://en.wikipedia.org/wiki/STUN" target="_blank" rel="noreferrer"><abbr title="Session Traversal Utilities for NAT">STUN</abbr></a>'),
 			_('To detect the public IPv4 address for unrestricted full-cone/one-to-one NATs'));
 
 		o = s.taboption('advanced', form.Value, 'stun_host', _('STUN host'));
-		o.depends('use_stun', '1');
 		o.datatype = 'host';
+		o.depends('use_stun', '1');
+		o.retain = true;
 
 		o = s.taboption('advanced', form.Value, 'stun_port', _('STUN port'));
-		o.depends('use_stun', '1');
 		o.datatype = 'port';
 		o.placeholder = '3478';
+		o.depends('use_stun', '1');
+		o.retain = true;
 
 		o = s.taboption('advanced', form.Flag, 'secure_mode', _('Enable secure mode'),
 			_('Allow adding port maps for requesting IP addresses only'));
 		o.default = '1';
 		o.depends('enable_upnp', '1');
+		o.retain = true;
+
+		s.taboption('advanced', form.Flag, 'ipv6_disable', _('Disable IPv6 mapping'));
 
 		o = s.taboption('advanced', form.Value, 'notify_interval', _('Notify interval'),
-			_('A 900s interval will result in %s notifications with the minimum max-age of 1800s', 'A 900s interval will result in %s (%s = SSDP) notifications with the minimum max-age of 1800s')
-				.format('<abbr title="Simple Service Discovery Protocol">SSDP</abbr>'));
-		o.datatype = 'uinteger';
+			_('A 900 s interval sends %s announcements with the min. %s header',
+				'A 900 s interval sends %s (%s = SSDP) announcements with the min. %s (%s = Cache-Control: max-age=1800) header')
+			.format('<abbr title="Simple Service Discovery Protocol">SSDP</abbr>', '<code>Cache-Control: max-age=1800</code>'));
+		o.datatype = 'min(900)';
 		o.placeholder = '900';
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
 		o = s.taboption('advanced', form.Value, 'port', _('SOAP/HTTP port'));
 		o.datatype = 'port';
 		o.placeholder = '5000';
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'presentation_url', _('Presentation URL'),
-			_('Report custom router web interface (presentation) URL'));
+		o = s.taboption('advanced', form.Value, 'presentation_url', _('Router/presentation URL'),
+			_('Report custom router web interface URL'));
 		o.placeholder = 'http://192.168.1.1/';
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
 		o = s.taboption('advanced', form.Value, 'uuid', _('Device UUID'));
-		o.depends('enable_upnp', '1');
+		// o.depends('enable_upnp', '1');
+		o.depends('to-disable-as-rarely-used', '1');
+		o.retain = true;
 
 		o = s.taboption('advanced', form.Value, 'model_number', _('Announced model number'));
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
 		o = s.taboption('advanced', form.Value, 'serial_number', _('Announced serial number'));
 		o.depends('enable_upnp', '1');
+		o.retain = true;
 
 		o = s.taboption('advanced', form.Flag, 'system_uptime', _('Report system instead of service uptime'));
 		o.default = '1';
-		o.depends('enable_upnp', '1');
+		o.depends('to-disable-as-rarely-used', '1');
+		o.retain = true;
 
 		s.taboption('advanced', form.Flag, 'log_output', _('Enable additional logging'),
 			_('Puts extra debugging information into the system log'));
 
 		o = s.taboption('advanced', form.Value, 'upnp_lease_file', _('Service lease file'));
-		o.placeholder = '/var/run/miniupnpd.leases';
+		o.depends('to-disable-as-rarely-used', '1');
+		o.retain = true;
 
 		s = m.section(form.GridSection, 'perm_rule', _('Service Access Control List'),
 			_('ACL specify which client addresses and ports can be mapped, IPv6 always allowed.'));
-		s.sortable = true;
 		s.anonymous = true;
 		s.addremove = true;
+		s.sortable = true;
 
 		s.option(form.Value, 'comment', _('Comment'));
 
-		o = s.option(form.Value, 'int_addr', _('Client Address'));
+		o = s.option(form.Value, 'int_addr', _('IP address'));
 		o.datatype = 'ip4addr';
 		o.placeholder = '0.0.0.0/0';
 
-		o = s.option(form.Value, 'int_ports', _('Client Port'));
+		o = s.option(form.Value, 'int_ports', _('Port'));
 		o.datatype = 'portrange';
 		o.placeholder = '1-65535';
 
-		o = s.option(form.Value, 'ext_ports', _('External Port'));
+		o = s.option(form.Value, 'ext_ports', _('External port'));
 		o.datatype = 'portrange';
 		o.placeholder = '1-65535';
 
diff --git a/applications/luci-app-upnp/root/usr/share/rpcd/acl.d/luci-app-upnp.json b/applications/luci-app-upnp/root/usr/share/rpcd/acl.d/luci-app-upnp.json
index 4611f501692a..40cb4c17eac0 100644
--- a/applications/luci-app-upnp/root/usr/share/rpcd/acl.d/luci-app-upnp.json
+++ b/applications/luci-app-upnp/root/usr/share/rpcd/acl.d/luci-app-upnp.json
@@ -1,6 +1,6 @@
 {
 	"luci-app-upnp": {
-		"description": "Grant access to UPnP IGD & PCP/NAT-PMP",
+		"description": "Grant access to UPnP IGD & PCP",
 		"read": {
 			"ubus": {
 				"luci.upnp": [ "get_status" ],
diff --git a/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp b/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
index f72b57b7ba84..3249b9e32ea2 100644
--- a/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
+++ b/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
@@ -1,3 +1,5 @@
+#!/usr/bin/env ucode
+
 // Copyright 2022 Jo-Philipp Wich <jo@mein.io>
 // Licensed to the public under the Apache License 2.0.
 
@@ -7,47 +9,43 @@ import { access, open, popen } from 'fs';
 import { connect } from 'ubus';
 import { cursor } from 'uci';
 
-// Establish ubus connection persistently outside of the call handler scope to
-// prevent premature GC'ing. Can be moved into `get_status` callback once
-// https://github.com/jow-/ucode/commit/a58fe4709f661b5f28e26701ea8638efccf5aeb6
-// is merged.
-const ubus = connect();
+const uci = cursor();
+const leasefilepath = uci.get('upnpd', 'config', 'upnp_lease_file') || '/run/miniupnpd.leases';
 
 const methods = {
 	get_status: {
 		call: function(req) {
-			const uci = cursor();
-
+			const ubus = connect();
 			const rules = [];
 			const leases = [];
-
-			const leasefile = open(uci.get('upnpd', 'config', 'upnp_lease_file'), 'r');
-
+			const leasefile = open(leasefilepath, 'r');
 			if (leasefile) {
 				for (let line = leasefile.read('line'); length(line); line = leasefile.read('line')) {
 					const record = split(line, ':', 6);
-
 					if (length(record) == 6) {
+						let descr = trim(record[5]);
+						let m = match(descr, /^PCP [A-Z]+ ([0-9a-f]{24})$/);
+						if (m) descr = 'PCP (nonce ' + m[1] + ')';
+						else if (match(descr, /^NAT-PMP \d+ \w+$/)) descr = 'NAT-PMP';
+						else if (match(descr, /^IGD2 pinhole$/)) descr = 'UPnP IGDv2 IPv6';
+						else if (!match(descr, /^UPnP IGD/)) descr = 'UPnP IGD / ' + descr;
 						push(leases, {
 							proto: uc(record[0]),
 							extport: +record[1],
 							intaddr: arrtoip(iptoarr(record[2])),
 							intport: +record[3],
 							expires: record[4] - timelocal(localtime()),
-							description: trim(record[5])
+							descr: descr
 						});
 					}
 				}
-
 				leasefile.close();
 			}
 
 			const ipt = popen('iptables --line-numbers -t nat -xnvL MINIUPNPD 2>/dev/null');
-
 			if (ipt) {
 				for (let line = ipt.read('line'); length(line); line = ipt.read('line')) {
 					let m = match(line, /^([0-9]+)\s+([a-z]+).+dpt:([0-9]+) to:(\S+):([0-9]+)/);
-
 					if (m) {
 						push(rules, {
 							num: m[1],
@@ -59,16 +57,13 @@ const methods = {
 						});
 					}
 				}
-
 				ipt.close();
 			}
 
 			const nft = popen('nft --handle list chain inet fw4 upnp_prerouting 2>/dev/null');
-
 			if (nft) {
 				for (let line = nft.read('line'), num = 1; length(line); line = nft.read('line')) {
-					let m = match(line, /^\t\tiif ".+" @nh,72,8 (0x6|0x11) th dport ([0-9]+) dnat ip to ([0-9.]+):([0-9]+)/);
-
+					let m = match(line, /^\t\tiif ".+" @nh,72,8 (0x6|0x11) th dport ([0-9]+).* dnat ip to ([0-9.]+):([0-9]+)/);
 					if (m) {
 						push(rules, {
 							num: `${num}`,
@@ -78,11 +73,9 @@ const methods = {
 							intport: +m[4],
 							descr: ''
 						});
-
 						num++;
 					}
 				}
-
 				nft.close();
 			}
 
@@ -90,16 +83,14 @@ const methods = {
 				for (let rule in rules) {
 					for (let lease in leases) {
 						if (lease.proto == rule.proto &&
-						    lease.intaddr == rule.intaddr &&
-						    lease.intport == rule.intport &&
-						    lease.extport == rule.extport)
-						{
-							rule.descr = lease.description;
+							lease.intaddr == rule.intaddr &&
+							lease.intport == rule.intport &&
+							lease.extport == rule.extport) {
+							rule.descr = lease.descr;
 							rule.expires = lease.expires;
 							break;
 						}
 					}
-
 					for (let mac, hint in host_hints) {
 						if (rule.intaddr in hint.ipaddrs) {
 							rule.host_hint = hint.name;
@@ -107,7 +98,6 @@ const methods = {
 						}
 					}
 				}
-
 				req.reply({ rules });
 			});
 		}
@@ -117,19 +107,13 @@ const methods = {
 		args: { token: 'token' },
 		call: function(req) {
 			const idx = +req.args?.token;
-
 			if (idx > 0) {
-				const uci = cursor();
-				const leasefile = uci.get('upnpd', 'config', 'upnp_lease_file');
-
-				if (access(leasefile)) {
-					system(['sed', '-i', '-e', `${idx}d`, leasefile]);
+				if (access(leasefilepath)) {
+					system(['sed', '-i', '-e', `${idx}d`, leasefilepath]);
 					system(['/etc/init.d/miniupnpd', 'restart']);
 				}
-
 				return { result: 'OK' };
 			}
-
 			return { result: 'Bad request' };
 		}
 	}

From b9dbf860350e8eb09fdabc4d9d8e4def2c4f7b59 Mon Sep 17 00:00:00 2001
From: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
Date: Mon, 29 Dec 2025 00:00:00 +0000
Subject: [PATCH 2/6] luci-app-upnp: add `UPnP IGD Adjustments` tab

And rearrange as many options

(to merge with prior)

Signed-off-by: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
---
 .../luci-static/resources/view/upnp/upnp.js   | 69 ++++++++++---------
 1 file changed, 35 insertions(+), 34 deletions(-)

diff --git a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
index 5b664db4c5be..7d515e65248c 100644
--- a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
+++ b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
@@ -144,6 +144,7 @@ return view.extend({
 		s.addremove = false;
 		s.tab('setup', _('Service Setup'));
 		s.tab('advanced', _('Advanced Settings'));
+		s.tab('igd', _('UPnP IGD Adjustments'));
 
 		o = s.taboption('setup', form.Flag, 'enabled', _('Start service'),
 			_('Start the autonomous port mapping service'));
@@ -162,16 +163,6 @@ return view.extend({
 		o.depends('enable_upnp', '1');
 		o.retain = true;
 
-		o = s.taboption('setup', form.Value, 'download', _('Download speed'),
-			_('Report maximum download speed in kByte/s'));
-		o.depends('enable_upnp', '1');
-		o.retain = true;
-
-		o = s.taboption('setup', form.Value, 'upload', _('Upload speed'),
-			_('Report maximum upload speed in kByte/s'));
-		o.depends('enable_upnp', '1');
-		o.retain = true;
-
 		s.taboption('advanced', form.Flag, 'use_stun', _('Use %s', 'Use %s (%s = STUN)')
 			.format('<a href="https://en.wikipedia.org/wiki/STUN" target="_blank" rel="noreferrer"><abbr title="Session Traversal Utilities for NAT">STUN</abbr></a>'),
 			_('To detect the public IPv4 address for unrestricted full-cone/one-to-one NATs'));
@@ -195,50 +186,60 @@ return view.extend({
 
 		s.taboption('advanced', form.Flag, 'ipv6_disable', _('Disable IPv6 mapping'));
 
-		o = s.taboption('advanced', form.Value, 'notify_interval', _('Notify interval'),
-			_('A 900 s interval sends %s announcements with the min. %s header',
-				'A 900 s interval sends %s (%s = SSDP) announcements with the min. %s (%s = Cache-Control: max-age=1800) header')
-			.format('<abbr title="Simple Service Discovery Protocol">SSDP</abbr>', '<code>Cache-Control: max-age=1800</code>'));
-		o.datatype = 'min(900)';
-		o.placeholder = '900';
-		o.depends('enable_upnp', '1');
+		o = s.taboption('advanced', form.Flag, 'system_uptime', _('Report system instead of service uptime'));
+		o.default = '1';
+		o.depends('to-disable-as-rarely-used', '1');
 		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'port', _('SOAP/HTTP port'));
-		o.datatype = 'port';
-		o.placeholder = '5000';
+		s.taboption('advanced', form.Flag, 'log_output', _('Enable additional logging'),
+			_('Puts extra debugging information into the system log'));
+
+		o = s.taboption('advanced', form.Value, 'upnp_lease_file', _('Service lease file'));
+		o.depends('to-disable-as-rarely-used', '1');
+		o.retain = true;
+
+		o = s.taboption('igd', form.Value, 'download', _('Download speed'),
+			_('Report maximum download speed in kByte/s'));
 		o.depends('enable_upnp', '1');
 		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'presentation_url', _('Router/presentation URL'),
-			_('Report custom router web interface URL'));
-		o.placeholder = 'http://192.168.1.1/';
+		o = s.taboption('igd', form.Value, 'upload', _('Upload speed'),
+			_('Report maximum upload speed in kByte/s'));
 		o.depends('enable_upnp', '1');
 		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'uuid', _('Device UUID'));
-		// o.depends('enable_upnp', '1');
-		o.depends('to-disable-as-rarely-used', '1');
+		o = s.taboption('igd', form.Value, 'model_number', _('Announced model number'));
+		o.depends('enable_upnp', '1');
 		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'model_number', _('Announced model number'));
+		o = s.taboption('igd', form.Value, 'serial_number', _('Announced serial number'));
 		o.depends('enable_upnp', '1');
 		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'serial_number', _('Announced serial number'));
+		o = s.taboption('igd', form.Value, 'presentation_url', _('Router/presentation URL'),
+			_('Report custom router web interface URL'));
+		o.placeholder = 'http://192.168.1.1/';
 		o.depends('enable_upnp', '1');
 		o.retain = true;
 
-		o = s.taboption('advanced', form.Flag, 'system_uptime', _('Report system instead of service uptime'));
-		o.default = '1';
+		o = s.taboption('igd', form.Value, 'uuid', _('Device UUID'));
+		// o.depends('enable_upnp', '1');
 		o.depends('to-disable-as-rarely-used', '1');
 		o.retain = true;
 
-		s.taboption('advanced', form.Flag, 'log_output', _('Enable additional logging'),
-			_('Puts extra debugging information into the system log'));
+		o = s.taboption('igd', form.Value, 'port', _('SOAP/HTTP port'));
+		o.datatype = 'port';
+		o.placeholder = '5000';
+		o.depends('enable_upnp', '1');
+		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'upnp_lease_file', _('Service lease file'));
-		o.depends('to-disable-as-rarely-used', '1');
+		o = s.taboption('igd', form.Value, 'notify_interval', _('Notify interval'),
+			_('A 900 s interval sends %s announcements with the min. %s header',
+				'A 900 s interval sends %s (%s = SSDP) announcements with the min. %s (%s = Cache-Control: max-age=1800) header')
+			.format('<abbr title="Simple Service Discovery Protocol">SSDP</abbr>', '<code>Cache-Control: max-age=1800</code>'));
+		o.datatype = 'min(900)';
+		o.placeholder = '900';
+		o.depends('enable_upnp', '1');
 		o.retain = true;
 
 		s = m.section(form.GridSection, 'perm_rule', _('Service Access Control List'),

From 0b1b8e1bd8c648a3a3c831e6487a6535df5a8490 Mon Sep 17 00:00:00 2001
From: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
Date: Mon, 29 Dec 2025 00:00:00 +0000
Subject: [PATCH 3/6] luci-app-upnp: revision and adapt to updated package
 options

The following settings UCI options been added or changed, and the
previous options are migrated on updating:

- Only display `Active Port Maps` if the service is enabled and `Access
  Control List` if it is used
- Enabled protocols (`enabled_protocols`): Combined UI option added
- Allow CGNAT/STUN (`allow_cgnat`): Allow new option for IPv4 CGNAT use
  (allow filtered), and updated help with newer wording of RFC 5780
- STUN server (`stun_host`): Allow port inclusion
- STUN port: Removed, as now accepted in STUN server
- Override external IPv4 (`external_ip`): UI option added for CGNAT use
- Allow third-party mapping (`allow_third_party_mapping`): Inverted from
  secure mode and optionally extended to PCP
- Log output level (`log_output`): Allow info log level, and reworded
- UPnP IGD compatibility (`upnp_igd_compat`): Reworded/extensible
- Download/upload speed (`download_kbps`/`upload_kbps`): In kbit/s and
  datatype set, now, interface link speed by default
- Router/friendly name (`friendly_name`): UI option added to set name
  displayed in Windows Explorer, model/serial number removed
- Enabled Networks / Access Control (`internal_network`): Section added
  to choose the enabled networks and their access control. By:
  - Internal network (`interface`): UI option added to select the
    local/internal (LAN) network interface to be enabled
  - Access control preset (`access_preset`): UI option added to select a
    preset for all devices on this network
  - Accept extra ports (`accept_ports`): UI option added to accept
    additional ports or port ranges on this network
  - Reject extra ports (`reject_ports`): UI option added to reject ports
    on the network regardless of other settings
  - Check ACL before (`acl_before`): UI option added to decide if the
    ACL entries should be checked before a preset
- Improve introduction text slightly

More details on changed options can be found in the dependent package PR

Depends on: https://redirect.github.com/openwrt/packages/pull/24988

Signed-off-by: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
---
 applications/luci-app-upnp/Makefile           |   2 +-
 .../luci-static/resources/view/upnp/upnp.js   | 208 +++++++++++++-----
 .../root/usr/share/rpcd/ucode/luci.upnp       |   2 +-
 3 files changed, 159 insertions(+), 53 deletions(-)

diff --git a/applications/luci-app-upnp/Makefile b/applications/luci-app-upnp/Makefile
index 80e668764d96..5ab5cc58275b 100644
--- a/applications/luci-app-upnp/Makefile
+++ b/applications/luci-app-upnp/Makefile
@@ -6,7 +6,7 @@
 
 include $(TOPDIR)/rules.mk
 
-LUCI_TITLE:=UPnP IGD & PCP/NAT-PMP configuration module | Universal Plug and Play
+LUCI_TITLE:=LuCI support for UPnP IGD & PCP/NAT-PMP service | formerly: Universal Plug and Play
 LUCI_DEPENDS:=+luci-base +miniupnpd +rpcd-mod-ucode
 
 PKG_LICENSE:=Apache-2.0
diff --git a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
index 7d515e65248c..d44c30abc66b 100644
--- a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
+++ b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
@@ -6,6 +6,7 @@
 'require ui';
 'require rpc';
 'require form';
+'require tools.widgets as widgets';
 
 const callInitAction = rpc.declare({
 	object: 'luci',
@@ -87,8 +88,8 @@ return view.extend({
 			'<a href="https://en.wikipedia.org/wiki/Port_Control_Protocol" target="_blank" rel="noreferrer"><abbr title="Port Control Protocol">PCP</abbr></a>',
 			'<a href="https://en.wikipedia.org/wiki/NAT_Port_Mapping_Protocol" target="_blank" rel="noreferrer"><abbr title="NAT Port Mapping Protocol">NAT-PMP</abbr></a>');
 		m = new form.Map('upnpd', _('UPnP IGD & PCP/NAT-PMP Service'),
-			_('The %s protocols allow clients on the local network to configure port maps/forwards on the router autonomously.',
-				'The %s (%s = UPnP IGD & PCP/NAT-PMP) protocols allow clients on the local network to configure port maps/forwards on the router autonomously.')
+			_('The %s protocols/service enable allowed devices on local networks to autonomously set up port maps (forwards) on this router.',
+				'The %s (%s = UPnP IGD & PCP/NAT-PMP) protocols/service enable allowed devices on local networks to autonomously set up port maps (forwards) on this router.')
 			.format(protocols)
 		);
 		if (!uci.get('upnpd', 'config')) {
@@ -100,6 +101,7 @@ return view.extend({
 		}
 
 		s = m.section(form.GridSection, '_active_rules');
+		s.disable = uci.get('upnpd', 'config', 'enabled') == '0';
 
 		s.render = L.bind(function(view, section_id) {
 			const table = E('table', { 'class': 'table cbi-section-table', 'id': 'upnp_status_table' }, [
@@ -150,39 +152,49 @@ return view.extend({
 			_('Start the autonomous port mapping service'));
 		o.rmempty = false;
 
-		o = s.taboption('setup', form.Flag, 'enable_upnp', _('Enable UPnP IGD protocol'));
-		o.default = '1';
-
-		o = s.taboption('setup', form.Flag, 'enable_natpmp', _('Enable PCP/NAT-PMP protocols'));
-		o.default = '1';
-
-		o = s.taboption('setup', form.Flag, 'igdv1', _('UPnP IGDv1 compatibility mode'),
-			_('Advertise as IGDv1 (IPv4 only) device instead of IGDv2'));
-		o.default = '1';
-		o.rmempty = false;
-		o.depends('enable_upnp', '1');
+		o = s.taboption('setup', form.ListValue, 'enabled_protocols', _('Enabled protocols'));
+		o.value('all', _('All protocols'));
+		o.value('upnp-igd', _('UPnP IGD'));
+		o.value('pcp+nat-pmp', _('PCP and NAT-PMP'));
+		o.default = 'all';
+		o.widget = 'radio';
+
+		o = s.taboption('setup', form.ListValue, 'upnp_igd_compat', _('UPnP IGD compatibility'),
+			_('Set compatibility mode (act as device) to workaround IGDv2-incompatible clients; %s only work with %s (or) Act/emulate as a specific/different device to workaround/support/handle/bypass/assist/mitigate IGDv2-incompatible clients (alternative text welcome)').format('PlayStation, CoD', 'IGDv1'));
+		o.value('igdv1', _('IGDv1 (IPv4 only)'));
+		o.value('igdv2', _('IGDv2 (with workarounds)'));
+		o.depends('enabled_protocols', 'upnp-igd');
+		o.depends('enabled_protocols', 'all');
 		o.retain = true;
 
-		s.taboption('advanced', form.Flag, 'use_stun', _('Use %s', 'Use %s (%s = STUN)')
-			.format('<a href="https://en.wikipedia.org/wiki/STUN" target="_blank" rel="noreferrer"><abbr title="Session Traversal Utilities for NAT">STUN</abbr></a>'),
-			_('To detect the public IPv4 address for unrestricted full-cone/one-to-one NATs'));
-
-		o = s.taboption('advanced', form.Value, 'stun_host', _('STUN host'));
-		o.datatype = 'host';
-		o.depends('use_stun', '1');
+		o = s.taboption('advanced', form.RichListValue, 'allow_cgnat', _('Allow %s/%s', 'Allow %s/%s (%s = CGNAT, %s = STUN)')
+			.format('<a href="https://en.wikipedia.org/wiki/Carrier-grade_NAT" target="_blank" rel="noreferrer"><abbr title="Carrier-grade NAT">CGNAT</abbr></a>',
+				'<a href="https://en.wikipedia.org/wiki/STUN" target="_blank" rel="noreferrer"><abbr title="Session Traversal Utilities for NAT">STUN</abbr></a>'),
+			_('Allow use of unrestricted endpoint-independent (1:1) CGNATs and detect the public IPv4'));
+		o.value('', _('Disabled'), _('Override public/external IPv4 address to prevent a private'));
+		o.value('1', _('Enabled'), _('Filtering test currently requires an extra firewall rule'));
+		o.value('allow-filtered', _('Enabled') + ' (' + _('allow filtered') + ')', _('Allow filtered IPv4 CGNAT test result'));
+		o.optional = true;
+
+		o = s.taboption('advanced', form.Value, 'stun_host', _('STUN server'));
+		o.datatype = 'or(hostname,hostport,ip4addr("nomask"))';
+		o.placeholder = 'stun.nextcloud.com';
+		o.depends('allow_cgnat', '1');
+		o.depends('allow_cgnat', 'allow-filtered');
 		o.retain = true;
 
-		o = s.taboption('advanced', form.Value, 'stun_port', _('STUN port'));
-		o.datatype = 'port';
-		o.placeholder = '3478';
-		o.depends('use_stun', '1');
-		o.retain = true;
+		o = s.taboption('advanced', form.Value, 'external_ip', _('Override external IPv4'),
+			_('Report custom public/external (WAN) IPv4 address'));
+		o.datatype = 'ip4addr("nomask")';
+		o.placeholder = '(203.1.2.3)';
+		o.depends('allow_cgnat', '');
 
-		o = s.taboption('advanced', form.Flag, 'secure_mode', _('Enable secure mode'),
-			_('Allow adding port maps for requesting IP addresses only'));
-		o.default = '1';
-		o.depends('enable_upnp', '1');
-		o.retain = true;
+		o = s.taboption('advanced', form.ListValue, 'allow_third_party_mapping', _('Allow third-party mapping'),
+			_('Allow adding port maps for non-requesting IP addresses'));
+		o.value('', _('Disabled'));
+		o.value('1', _('Enabled'));
+		o.value('upnp-igd', _('Enabled') + ' (' + _('UPnP IGD only') + ')');
+		o.value('pcp', _('Enabled') + ' (' + _('PCP only') + ')');
 
 		s.taboption('advanced', form.Flag, 'ipv6_disable', _('Disable IPv6 mapping'));
 
@@ -191,46 +203,69 @@ return view.extend({
 		o.depends('to-disable-as-rarely-used', '1');
 		o.retain = true;
 
-		s.taboption('advanced', form.Flag, 'log_output', _('Enable additional logging'),
-			_('Puts extra debugging information into the system log'));
+		o = s.taboption('advanced', form.ListValue, 'log_output', _('Log output level'));
+		o.value('default', _('Default'));
+		o.value('info', _('Info'));
+		o.value('debug', _('Debug'));
+		o.default = 'default';
+		o.widget = 'radio';
 
-		o = s.taboption('advanced', form.Value, 'upnp_lease_file', _('Service lease file'));
+		o = s.taboption('advanced', form.Value, 'lease_file', _('Service lease file'));
 		o.depends('to-disable-as-rarely-used', '1');
 		o.retain = true;
 
-		o = s.taboption('igd', form.Value, 'download', _('Download speed'),
-			_('Report maximum download speed in kByte/s'));
-		o.depends('enable_upnp', '1');
+		o = s.taboption('igd', form.Value, 'download_kbps', _('Download speed'),
+			_('Report maximum connection speed in kbit/s'));
+		o.datatype = 'uinteger';
+		o.placeholder = _('Default interface link speed');
+		o.depends('enabled_protocols', 'upnp-igd');
+		o.depends('enabled_protocols', 'all');
+		o.retain = true;
+
+		o = s.taboption('igd', form.Value, 'upload_kbps', _('Upload speed'),
+			_('Report maximum connection speed in kbit/s'));
+		o.datatype = 'uinteger';
+		o.placeholder = _('Default interface link speed');
+		o.depends('enabled_protocols', 'upnp-igd');
+		o.depends('enabled_protocols', 'all');
 		o.retain = true;
 
-		o = s.taboption('igd', form.Value, 'upload', _('Upload speed'),
-			_('Report maximum upload speed in kByte/s'));
-		o.depends('enable_upnp', '1');
+		o = s.taboption('igd', form.Value, 'friendly_name', _('Router/friendly name'));
+		o.placeholder = 'OpenWrt UPnP IGD & PCP';
+		o.depends('enabled_protocols', 'upnp-igd');
+		o.depends('enabled_protocols', 'all');
 		o.retain = true;
 
 		o = s.taboption('igd', form.Value, 'model_number', _('Announced model number'));
-		o.depends('enable_upnp', '1');
+		// o.depends('enabled_protocols', 'upnp-igd');
+		// o.depends('enabled_protocols', 'all');
+		o.depends('to-disable-as-rarely-used', '1');
 		o.retain = true;
 
 		o = s.taboption('igd', form.Value, 'serial_number', _('Announced serial number'));
-		o.depends('enable_upnp', '1');
+		// o.depends('enabled_protocols', 'upnp-igd');
+		// o.depends('enabled_protocols', 'all');
+		o.depends('to-disable-as-rarely-used', '1');
 		o.retain = true;
 
 		o = s.taboption('igd', form.Value, 'presentation_url', _('Router/presentation URL'),
 			_('Report custom router web interface URL'));
 		o.placeholder = 'http://192.168.1.1/';
-		o.depends('enable_upnp', '1');
+		o.depends('enabled_protocols', 'upnp-igd');
+		o.depends('enabled_protocols', 'all');
 		o.retain = true;
 
 		o = s.taboption('igd', form.Value, 'uuid', _('Device UUID'));
-		// o.depends('enable_upnp', '1');
+		// o.depends('enabled_protocols', 'upnp-igd');
+		// o.depends('enabled_protocols', 'all');
 		o.depends('to-disable-as-rarely-used', '1');
 		o.retain = true;
 
-		o = s.taboption('igd', form.Value, 'port', _('SOAP/HTTP port'));
+		o = s.taboption('igd', form.Value, 'http_port', _('SOAP/HTTP port'));
 		o.datatype = 'port';
 		o.placeholder = '5000';
-		o.depends('enable_upnp', '1');
+		o.depends('enabled_protocols', 'upnp-igd');
+		o.depends('enabled_protocols', 'all');
 		o.retain = true;
 
 		o = s.taboption('igd', form.Value, 'notify_interval', _('Notify interval'),
@@ -239,7 +274,68 @@ return view.extend({
 			.format('<abbr title="Simple Service Discovery Protocol">SSDP</abbr>', '<code>Cache-Control: max-age=1800</code>'));
 		o.datatype = 'min(900)';
 		o.placeholder = '900';
-		o.depends('enable_upnp', '1');
+		o.depends('enabled_protocols', 'upnp-igd');
+		o.depends('enabled_protocols', 'all');
+		o.retain = true;
+
+		s = m.section(form.GridSection, 'internal_network', _('Enabled Networks / Access Control'),
+			_('Select local/internal (LAN) network interfaces to enable the service for.') + ' ' +
+			_('Select an access control preset that defines which ports all devices on a network can map.') + ' ' +
+			_('Alternatively, add client-specific permissions using the access control list (ACL), which can extend/override a preset.') + ' ' +
+			_('IPv6 is currently always accepted unless disabled. (alternative text welcome)'));
+		s.anonymous = true;
+		s.addremove = true;
+		s.cloneable = true;
+		s.sortable = true;
+		s.nodescriptions = true;
+		s.modaltitle = _('UPnP IGD & PCP') + ' - ' + _('Edit network access control settings');
+
+		o = s.option(widgets.NetworkSelect, 'interface', _('Internal network'),
+			_('Select the local/internal (LAN) network interface to enable the service for'));
+		o.exclude = 'wan'; // wan6 should also be excluded
+		o.nocreate = true;
+		o.editable = true;
+		o.retain = true;
+		o.validate = function(section_id, value) {
+			let netcount = 0;
+			// Commented out, as it causes issues with cloning
+			//for (let ifnr = 0; uci.get('upnpd', `@internal_network[${ifnr}]`, 'interface'); ifnr++) {
+			//	if (uci.get('upnpd', `@internal_network[${ifnr}]`, 'interface') == value) netcount++;
+			//};
+			return (netcount >= 2 || value == '' || value == 'wan' || value == 'wan6') ? '' : true;
+		};
+
+		o = s.option(form.ListValue, 'access_preset', _('Access control preset'),
+			_('Select a preset for all devices on this network'));
+		o.value('none', _('None / accept listed ports only'));
+		o.value('accept-high-ports', _('Accept ports >=1024'));
+		o.value('accept-high-ports+web', _('Accept HTTP/HTTPS + ports >=1024'));
+		o.value('accept-high-ports+web+dns', _('Accept HTTP/HTTPS/DNS + ports >=1024'));
+		o.value('accept-all-ports', _('Accept all ports'));
+		o.editable = true;
+		o.retain = true;
+
+		o = s.option(form.Value, 'accept_ports', _('Accept extra ports'),
+			_('Accept these ports or port ranges on this network'));
+		o.depends({ access_preset: 'accept-all-ports', '!reverse': true });
+		o.retain = true;
+		o.validate = function(section_id, value) {
+			return value.search(/^[0-9 -]*$/) != -1 ? true : '';
+		};
+
+		o = s.option(form.Value, 'reject_ports', _('Reject extra ports'),
+			_('Reject unsafe/insecure/risky FTP/Telnet/DCE/NetBIOS/SMB/RDP ports on this network by default, before other settings'));
+		o.placeholder = '21 23 135 137-139 445 3389';
+		o.modalonly = true;
+		o.retain = true;
+		o.validate = function(section_id, value) {
+			return value.search(/^[0-9 -]*$/) != -1 ? true : '';
+		};
+
+		o = s.option(form.Flag, 'acl_before', _('Check ACL before'),
+			_('Whether the access control list (ACL) entries should be checked before a preset; can extend/override a preset') + '<br>' +
+			_('Sequence: 1. Reject extra ports, 2. ACL entries if enabled, 3. Preset ports, 4. Accept extra ports'));
+		o.editable = true;
 		o.retain = true;
 
 		s = m.section(form.GridSection, 'perm_rule', _('Service Access Control List'),
@@ -247,6 +343,14 @@ return view.extend({
 		s.anonymous = true;
 		s.addremove = true;
 		s.sortable = true;
+		// Preferably: ACL part of extra tab with depends for section as immediately, and network section part of service setup tab. Nice to have: Add button (+input) calls function and opens modal pre-filled
+		let acl_used = false;
+		for (let ifnr = 0; uci.get('upnpd', `@internal_network[${ifnr}]`, 'interface'); ifnr++) {
+			if (uci.get('upnpd', `@internal_network[${ifnr}]`, 'acl_before') == '1') {
+				acl_used = true;
+			}
+		}
+		s.disable = !acl_used;
 
 		s.option(form.Value, 'comment', _('Comment'));
 
@@ -267,11 +371,13 @@ return view.extend({
 		o.value('deny', _('Deny'));
 
 		return m.render().then(L.bind(function(m, nodes) {
-			poll.add(L.bind(function() {
-				return Promise.all([
-					callUpnpGetStatus()
-				]).then(L.bind(this.poll_status, this, nodes));
-			}, this), 5);
+			if (uci.get('upnpd', 'config', 'enabled') != '0') {
+				poll.add(L.bind(function() {
+					return Promise.all([
+						callUpnpGetStatus()
+					]).then(L.bind(this.poll_status, this, nodes));
+				}, this), 5);
+			}
 			return nodes;
 		}, this, m));
 	}
diff --git a/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp b/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
index 3249b9e32ea2..e9fa04e3f224 100644
--- a/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
+++ b/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
@@ -10,7 +10,7 @@ import { connect } from 'ubus';
 import { cursor } from 'uci';
 
 const uci = cursor();
-const leasefilepath = uci.get('upnpd', 'config', 'upnp_lease_file') || '/run/miniupnpd.leases';
+const leasefilepath = uci.get('upnpd', 'config', 'lease_file') || '/run/miniupnpd.leases';
 
 const methods = {
 	get_status: {

From 21439841e3ae0e25f3b73d87ec202377df302256 Mon Sep 17 00:00:00 2001
From: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
Date: Mon, 29 Dec 2025 00:00:00 +0000
Subject: [PATCH 4/6] luci-app-upnp: rename UCI section name to `settings`
 (v2.0)

Inspired/address copilot's PR review for a clearer config by rename UCI
section name `config` (v1.0) -> `settings` (v2.0), helps on migration
and to distinguish the updated config from the previous one easily

(to merge with prior)

Signed-off-by: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
---
 .../htdocs/luci-static/resources/view/upnp/upnp.js     | 10 +++++-----
 .../luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp  |  2 +-
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
index d44c30abc66b..0b305c32c294 100644
--- a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
+++ b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
@@ -92,16 +92,16 @@ return view.extend({
 				'The %s (%s = UPnP IGD & PCP/NAT-PMP) protocols/service enable allowed devices on local networks to autonomously set up port maps (forwards) on this router.')
 			.format(protocols)
 		);
-		if (!uci.get('upnpd', 'config')) {
+		if (!uci.get('upnpd', 'settings')) {
 			ui.addNotification(null, E('div', '<h4>' + _('No suitable configuration was found!') + '</h4><p>' +
-				_('No suitable (LuCI app %s) config found in %s. Related package update (daemon or LuCI app) may be missing.').format('v1.0', '<code>/etc/config/upnpd</code>') + '<br>' +
+				_('No suitable (LuCI app %s) config found in %s. Related package update (daemon or LuCI app) may be missing.').format('v2.0', '<code>/etc/config/upnpd</code>') + '<br>' +
 				_('Use the software package manager, update lists, and install the related update. Config is migrated on the daemon package update.') + '</p>' +
 				'<a class="btn" href="/cgi-bin/luci/admin/system/package-manager?query=UPnP%20IGD%20&%20PCP/NAT-PMP">' + _('Go to package manager…') + '</a>'), 'warning');
 			m.readonly = true;
 		}
 
 		s = m.section(form.GridSection, '_active_rules');
-		s.disable = uci.get('upnpd', 'config', 'enabled') == '0';
+		s.disable = uci.get('upnpd', 'settings', 'enabled') == '0';
 
 		s.render = L.bind(function(view, section_id) {
 			const table = E('table', { 'class': 'table cbi-section-table', 'id': 'upnp_status_table' }, [
@@ -142,7 +142,7 @@ return view.extend({
 			]);
 		}, o, this);
 
-		s = m.section(form.NamedSection, 'config', 'upnpd', _('Service Settings'));
+		s = m.section(form.NamedSection, 'settings', 'upnpd', _('Service Settings'));
 		s.addremove = false;
 		s.tab('setup', _('Service Setup'));
 		s.tab('advanced', _('Advanced Settings'));
@@ -371,7 +371,7 @@ return view.extend({
 		o.value('deny', _('Deny'));
 
 		return m.render().then(L.bind(function(m, nodes) {
-			if (uci.get('upnpd', 'config', 'enabled') != '0') {
+			if (uci.get('upnpd', 'settings', 'enabled') != '0') {
 				poll.add(L.bind(function() {
 					return Promise.all([
 						callUpnpGetStatus()
diff --git a/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp b/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
index e9fa04e3f224..31bd39a9e7ff 100644
--- a/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
+++ b/applications/luci-app-upnp/root/usr/share/rpcd/ucode/luci.upnp
@@ -10,7 +10,7 @@ import { connect } from 'ubus';
 import { cursor } from 'uci';
 
 const uci = cursor();
-const leasefilepath = uci.get('upnpd', 'config', 'lease_file') || '/run/miniupnpd.leases';
+const leasefilepath = uci.get('upnpd', 'settings', 'lease_file') || '/run/miniupnpd.leases';
 
 const methods = {
 	get_status: {

From 951e51cb8232de2e4f26352a8cd37d94c7acc838 Mon Sep 17 00:00:00 2001
From: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
Date: Mon, 29 Dec 2025 00:00:00 +0000
Subject: [PATCH 5/6] luci-app-upnp: add second CGNAT UCI option

Alternative option to STUN allow-filtered. As requested by AquanJSW, to
test with Tailscale. Also adds the required daemon fix. No STUN public
IPv4 address detection; issues with multiple clients, e.g. PCP/NAT-PMP

(proposed for inclusion, to merge with prior)

Signed-off-by: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
---
 .../luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js | 1 +
 1 file changed, 1 insertion(+)

diff --git a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
index 0b305c32c294..cb0a87e78adf 100644
--- a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
+++ b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
@@ -174,6 +174,7 @@ return view.extend({
 		o.value('', _('Disabled'), _('Override public/external IPv4 address to prevent a private'));
 		o.value('1', _('Enabled'), _('Filtering test currently requires an extra firewall rule'));
 		o.value('allow-filtered', _('Enabled') + ' (' + _('allow filtered') + ')', _('Allow filtered IPv4 CGNAT test result'));
+		o.value('allow-private-ext-ipv4', _('Enabled') + ' (' + _('avoid, allow private external IPv4') + ')', _('No STUN public IPv4 address detection; issues with multiple clients'));
 		o.optional = true;
 
 		o = s.taboption('advanced', form.Value, 'stun_host', _('STUN server'));

From f94173b2dd3b971364b8292999065edf9d4b0957 Mon Sep 17 00:00:00 2001
From: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
Date: Mon, 29 Dec 2025 00:00:00 +0000
Subject: [PATCH 6/6] luci-app-upnp: update ACL options, migrate section

- Ignorable and cloneable ACL entries, always translated `Action`
- Improve UI with direct editability, clearer help wording, and rename
  to `Access Control List`
- Note that the ACL is now rejected by default, if it is alone used,
  with no preset or listed accepted ports. Add (ignored) ACL template
  entries on migration
- Migrate ACL entries to the new section name `acl_entry`
- The following ACL UCI options been added or changed, and the previous
  options are migrated on updating:

acl_entry UCI options       | Change                     | Previous name
----------------------------|----------------------------|--------------
action                      | New/updated values     (1) |
int_port                    | Remove colon separator (2) | int_ports
ext_port                    | Remove colon separator (2) | ext_ports
descr_filter                | New option             (3) |

1. Allow ignore, and update action option to use the nftables terms
   (allow/deny -> accept/reject). To avoid adding inverted actions when
   changing via LuCI, ensure any missing are set, as LuCI and UCI had
   not matching action defaults. Missing actions are now ignored/logged
2. Ensure that the hyphen (-) is only used as a port range separator by
   migration, as the colon (:) is not valid in LuCI
3. Add missing UCI option to set a regular expression to check for a
   UPnP IGD IPv4 port map description, and fix the current collision
   with the comment field which was not noticed due to a daemon bug
   https://redirect.github.com/openwrt/packages/pull/24495
   https://redirect.github.com/miniupnp/miniupnp/pull/853

- Refactoring by adding a more universal usable `is_port_or_range`
  function instead of `upnpd_get_port_range` and check if it has a valid
  range, and removes a shellcheck warning
- Rename `conf_rule_add` function to `upnpd_add_acl_entry`

(to merge with prior)

Signed-off-by: Self-Hosting-Group <selfhostinggroup-git+openwrt@shost.ing>
---
 .../luci-static/resources/view/upnp/upnp.js   | 38 ++++++++++++++-----
 1 file changed, 28 insertions(+), 10 deletions(-)

diff --git a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
index cb0a87e78adf..7849fc091903 100644
--- a/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
+++ b/applications/luci-app-upnp/htdocs/luci-static/resources/view/upnp/upnp.js
@@ -339,11 +339,14 @@ return view.extend({
 		o.editable = true;
 		o.retain = true;
 
-		s = m.section(form.GridSection, 'perm_rule', _('Service Access Control List'),
-			_('ACL specify which client addresses and ports can be mapped, IPv6 always allowed.'));
+		s = m.section(form.GridSection, 'acl_entry', _('Access Control List'),
+			_('The access control list (ACL) specifies which IP addresses and ports can be mapped.') + ' ' +
+			_('ACL entries are checked in order and rejected by default, with no preset. (should be part of extra tab)'));
 		s.anonymous = true;
 		s.addremove = true;
+		s.cloneable = true;
 		s.sortable = true;
+		s.modaltitle = _('UPnP IGD & PCP') + ' - ' + _('Edit access control list entry');
 		// Preferably: ACL part of extra tab with depends for section as immediately, and network section part of service setup tab. Nice to have: Add button (+input) calls function and opens modal pre-filled
 		let acl_used = false;
 		for (let ifnr = 0; uci.get('upnpd', `@internal_network[${ifnr}]`, 'interface'); ifnr++) {
@@ -353,23 +356,38 @@ return view.extend({
 		}
 		s.disable = !acl_used;
 
-		s.option(form.Value, 'comment', _('Comment'));
+		o = s.option(form.Value, 'comment', _('Comment'));
+		o.default = _('unspecified');
 
 		o = s.option(form.Value, 'int_addr', _('IP address'));
 		o.datatype = 'ip4addr';
-		o.placeholder = '0.0.0.0/0';
+		o.default = '0.0.0.0/0';
+		o.editable = true;
+		o.retain = true;
 
-		o = s.option(form.Value, 'int_ports', _('Port'));
+		o = s.option(form.Value, 'int_port', _('Port'));
 		o.datatype = 'portrange';
-		o.placeholder = '1-65535';
+		o.placeholder = '1-65535 (' + _('any port') + ')';
+		o.editable = true;
+		o.retain = true;
 
-		o = s.option(form.Value, 'ext_ports', _('External port'));
+		o = s.option(form.Value, 'ext_port', _('External port'));
 		o.datatype = 'portrange';
-		o.placeholder = '1-65535';
+		o.placeholder = '1-65535 (' + _('any port') + ')';
+		o.editable = true;
+		o.retain = true;
+
+		o = s.option(form.Value, 'descr_filter', _('Description filter'),
+			_('A regular expression to check for a UPnP IGD IPv4 port map description'));
+		o.placeholder = '^.*$ (' + _('any description') + ')';
+		o.modalonly = true;
 
 		o = s.option(form.ListValue, 'action', _('Action'));
-		o.value('allow', _('Allow'));
-		o.value('deny', _('Deny'));
+		o.value('accept', _('Accept'));
+		o.value('reject', _('Reject'));
+		o.value('ignore', _('Ignore'));
+		o.editable = true;
+		o.retain = true;
 
 		return m.render().then(L.bind(function(m, nodes) {
 			if (uci.get('upnpd', 'settings', 'enabled') != '0') {
